File: src/wbcore/clock.py
"""時刻の扱いの規約。ここ以外で「今」を取らない。

方針:
    保存
        時刻は必ずタイムゾーン付き。日中足の ``ts`` は UTC（Parquet に tz が残る）、
        SQLite の ``placed_at`` 等は UTC の ISO 8601（``+00:00`` 付き）。
        暦日（``date``）は取引所の日付であり時刻ではないので tz を持たない。
    演算・判定
        UTC。取引所の現地時刻が要る判断（発注時間帯・引けの前後）は、
        その場で :attr:`Market.timezone` に変換して比べる。
    表示
        既定は UTC。設定（``WBJP_TIMEZONE``）で JST などを指定できる。
        どちらの場合も**時間帯を必ず添える**（``2026-08-29 06:20 UTC`` /
        ``2026-08-29 15:20 JST``）。時間帯の無い時刻を画面に出さない。

naive（tz 無し）な datetime を受け取ったら UTC とみなす。ローカル時刻と
みなすと、cron のサーバーと開発機で結果が変わる。
"""

from __future__ import annotations

import datetime as dt
from zoneinfo import ZoneInfo

UTC = dt.UTC


def now_utc() -> dt.datetime:
    """現在時刻（UTC、tz 付き）。"""
    return dt.datetime.now(UTC)


def today_utc() -> dt.date:
    """UTC で見た今日。

    JST の深夜〜朝 9 時は UTC がまだ前日。「今日の足があるか」の判断を
    ローカル日付でやると、サーバーの時間帯次第で結果が変わる。
    """
    return now_utc().date()


def ensure_utc(moment: dt.datetime) -> dt.datetime:
    """tz 付き UTC に揃える。naive は UTC とみなす。"""
    if moment.tzinfo is None:
        return moment.replace(tzinfo=UTC)
    return moment.astimezone(UTC)


def zone(name: str | ZoneInfo | None) -> ZoneInfo:
    """時間帯の名前（``"Asia/Tokyo"`` / ``"UTC"``）を ZoneInfo に。None は UTC。"""
    if name is None:
        return ZoneInfo("UTC")
    if isinstance(name, ZoneInfo):
        return name
    try:
        return ZoneInfo(name)
    except Exception:
        raise ValueError(
            f"未知の時間帯: {name!r}（例: UTC / Asia/Tokyo / America/New_York）"
        ) from None


def to_zone(moment: dt.datetime, tz: str | ZoneInfo | None = None) -> dt.datetime:
    """表示用に時間帯を変える。naive は UTC とみなす。"""
    return ensure_utc(moment).astimezone(zone(tz))


def fmt(moment: dt.datetime, tz: str | ZoneInfo | None = None, *, seconds: bool = False) -> str:
    """時間帯の略号を添えて書く。``2026-08-29 15:20 JST`` / ``2026-08-29 06:20 UTC``。

    略号を持たない時間帯（``Asia/Tokyo`` は JST、``America/New_York`` は EDT/EST）
    は ``%Z`` が UTC からのずれ（``+09``）になる。どちらでも時間帯は分かる。
    """
    local = to_zone(moment, tz)
    pattern = "%Y-%m-%d %H:%M:%S %Z" if seconds else "%Y-%m-%d %H:%M %Z"
    return local.strftime(pattern)


def fmt_time(moment: dt.datetime, tz: str | ZoneInfo | None = None) -> str:
    """時刻だけ。``15:20 JST``。"""
    return to_zone(moment, tz).strftime("%H:%M %Z")


def fmt_iso(text: object, tz: str | ZoneInfo | None = None) -> str:
    """保存されている ISO 8601 の文字列（UTC）を表示用に変える。

    DB の ``placed_at`` 等をそのまま出すと ``2026-08-29T06:36:37+00:00`` で、
    時間帯を読み取れる人にしか意味が無い。設定の時間帯で ``2026-08-29 15:36:37 JST``
    にする。日時として読めない値はそのまま返す（日付や None を壊さない）。
    """
    if not isinstance(text, str):
        return str(text)
    try:
        moment = dt.datetime.fromisoformat(text.replace("Z", "+00:00"))
    except ValueError:
        return text
    if moment.tzinfo is None and "T" not in text:
        return text  # 日付だけ（"2026-08-29"）。時刻ではない
    return fmt(moment, tz, seconds=True)


def stamp_iso(tz: str | ZoneInfo | None = None) -> str:
    """ログ用の現在時刻。設定の時間帯で、オフセット付き ISO（``+09:00``）。"""
    return to_zone(now_utc(), tz).isoformat(timespec="microseconds")

File: pkg/wbcore/clock/clock.go
package clock

import (
	"fmt"
	"strings"
	"time"
)

var (
	// UTC は常に UTC ロケーションを表す。
	UTC = time.UTC

	// Tokyo は JST (Asia/Tokyo) ロケーションを表す。
	// tzdata の無い環境でも UTC に化けないよう、読めなければ固定の +09:00 にする。
	Tokyo = loadTokyo()

	// Now は現在時刻の出所。試験で時刻を固定したいときに差し替える
	// （NowUTC / NowJST / TodayUTC / TodayJST / StampISO はすべてここを通る）。
	Now = time.Now
)

// loadTokyo は Asia/Tokyo を読み、無ければ固定の JST（+09:00）を返す。
// 日本に夏時間は無いので、固定オフセットでも時刻は変わらない。
func loadTokyo() *time.Location {
	loc, err := time.LoadLocation("Asia/Tokyo")
	if err != nil || loc == nil {
		return time.FixedZone("JST", 9*3600)
	}
	return loc
}

// NowUTC は現在時刻を UTC で返す。
func NowUTC() time.Time {
	return Now().UTC()
}

// NowJST は現在時刻を JST で返す。
func NowJST() time.Time {
	return Now().In(Tokyo)
}

// TodayUTC は UTC で見た今日の日付（年月日のみ、時刻は 00:00:00 UTC）を返す。
func TodayUTC() time.Time {
	now := NowUTC()
	return time.Date(now.Year(), now.Month(), now.Day(), 0, 0, 0, 0, time.UTC)
}

// TodayJST は JST で見た今日の日付（年月日のみ、時刻は 00:00:00 JST）を返す。
// 日本の営業日を数えるコマンドはこちらを使う。TodayUTC は 09:00 JST までは前日になる。
func TodayJST() time.Time {
	now := NowJST()
	return time.Date(now.Year(), now.Month(), now.Day(), 0, 0, 0, 0, Tokyo)
}

// ParseDateJST は "YYYY-MM-DD" を JST の 00:00 として読む。
func ParseDateJST(s string) (time.Time, error) {
	return time.ParseInLocation("2006-01-02", s, Tokyo)
}

// EnsureUTC は時刻を UTC に揃える。タイムゾーン情報が未指定の場合は UTC とみなす。
func EnsureUTC(t time.Time) time.Time {
	return t.UTC()
}

// Zone はタイムゾーン名から time.Location を取得する。空文字や未知の場合は UTC またはエラーを返す。
func Zone(name string) (*time.Location, error) {
	if name == "" || name == "UTC" {
		return time.UTC, nil
	}
	loc, err := time.LoadLocation(name)
	if err != nil {
		return nil, fmt.Errorf("未知の時間帯: %q（例: UTC / Asia/Tokyo / America/New_York）: %w", name, err)
	}
	return loc, nil
}

// MustZone はタイムゾーン名から time.Location を取得し、失敗時は panic する。
func MustZone(name string) *time.Location {
	loc, err := Zone(name)
	if err != nil {
		panic(err)
	}
	return loc
}

// ToZone は時刻を指定したタイムゾーンに変換する。
func ToZone(t time.Time, loc *time.Location) time.Time {
	if loc == nil {
		loc = time.UTC
	}
	return t.In(loc)
}

// Fmt は時間帯の略号を添えてフォーマットする（例: "2026-08-29 15:20 JST" / "2026-08-29 06:20 UTC"）。
func Fmt(t time.Time, loc *time.Location, withSeconds bool) string {
	local := ToZone(t, loc)
	layout := "2006-01-02 15:04 MST"
	if withSeconds {
		layout = "2006-01-02 15:04:05 MST"
	}
	return local.Format(layout)
}

// FmtTime は時刻のみフォーマットする（例: "15:20 JST"）。
func FmtTime(t time.Time, loc *time.Location) string {
	local := ToZone(t, loc)
	return local.Format("15:04 MST")
}

// FmtISO は DB 等に保存されている ISO 8601 文字列（UTC）を表示用の指定タイムゾーン文字列に変換する。
// 日付のみ ("2026-08-29") の場合はそのまま返す。
func FmtISO(isoStr string, loc *time.Location) string {
	if !strings.Contains(isoStr, "T") && !strings.Contains(isoStr, " ") {
		return isoStr
	}
	parsed, err := time.Parse(time.RFC3339Nano, strings.Replace(isoStr, " ", "T", 1))
	if err != nil {
		parsed, err = time.Parse(time.RFC3339, isoStr)
		if err != nil {
			return isoStr
		}
	}
	return Fmt(parsed, loc, true)
}

// StampISO はログ用の現在時刻文字列（オフセット付き ISO 8601、マイクロ秒精度）を返す。
func StampISO(loc *time.Location) string {
	local := ToZone(NowUTC(), loc)
	return local.Format("2006-01-02T15:04:05.000000-07:00")
}
